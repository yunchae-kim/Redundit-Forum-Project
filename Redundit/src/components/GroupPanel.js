/* eslint-disable no-nested-ternary */
/* eslint-disable max-len */
import { Divider, Tag } from "@blueprintjs/core";
import { Link, useHistory } from "react-router-dom";
import { useEffect, useState } from "react";
import { getMember, createMember, deleteMember } from "../utility/axiosMember";
import { getGroupById } from "../utility/axiosGroup";
import { getGrouptags } from "../utility/axiosGrouptag";
import { getAllUsersInGroup } from "../utility/axiosUser";
import { getAllPostsInGroup } from "../utility/axiosPost";
import groupIcon from "../test_data/groupIcon.jpg";
import "../style/components/GroupPanel.css";
import httpClient from "../utility/auth.helper";
import { addNotification } from "../utility/axiosNotification";

const GroupPanel = ({ groupId }) => {
  const [group, setGroup] = useState({});
  const [fetched, setfetched] = useState(false);
  const [grouptags, setGrouptags] = useState([]);
  const [userNum, setUserNum] = useState(0);
  const [postNum, setPostNum] = useState(0);
  const [userId, setUserId] = useState(0);
  const [isJoined, setIsJoined] = useState(false);
  const [isPending, setIsPending] = useState(false);
  const [isAdmin, setIsAdmin] = useState(false);
  const [buttonText, setButtonText] = useState("");
  const history = useHistory();

  useState(() => {
    setfetched(true);
    getGroupById(groupId).then((groupFetched) => {
      setGroup(groupFetched);
    });
    getGrouptags(groupId).then((grouptagsFetched) => {
      setGrouptags(grouptagsFetched);
    });
    getAllUsersInGroup(groupId).then((usersInGroup) => {
      setUserNum(usersInGroup.length);
    });
    getAllPostsInGroup(groupId).then((postsInGroup) => {
      setPostNum(postsInGroup.length);
    });
  }, [fetched]);

  useEffect(() => {
    const userObj = httpClient.getCurrentUser();
    if (userObj) {
      const { iduser } = userObj;
      if (iduser) {
        setUserId(iduser);
        // member logic
        getMember(groupId, userId).then((memberFetched) => {
          if (memberFetched != null) {
            setIsJoined(true);
            if (memberFetched.pending === 1) {
              setIsPending(true);
            }
            if (memberFetched.admin === 1) {
              setIsAdmin(true);
            }
          }
        });
      }
    }
  }, [userId]);

  useEffect(() => {
    if (isJoined) {
      if (isAdmin) {
        setButtonText("Admin");
      } else if (isPending) {
        setButtonText("Pending");
      } else {
        setButtonText("Joined");
      }
    } else {
      setButtonText("Join");
    }
  }, [isJoined, isAdmin, isPending]);

  const hashTagsList = grouptags.map((h) => <Tag className="GP-tag" key={h.tag}>{h.tag}</Tag>);

  const unJoin = () => {
    if (isJoined && !isAdmin) {
      deleteMember(groupId, userId).then(() => {
        setIsJoined(!isJoined);
        setIsPending(false);
        history.push("/");
        history.go();
      });
    } else if (!isJoined) {
      createMember(groupId, userId, false, true).then(() => {
        setIsJoined(!isJoined);
        setIsPending(true);
        addNotification(group.creator, `Someone requested to join ${group.name}`).then(()=>{
          history.push({ pathname: `/group/${groupId}`, state: { groupId } });
          history.go();
        });
      });
    }
  };
  return (
    <div className="GroupPanel GroupPanel-small">
      <div className="GP-header">
        <img className="GP-groupIcon" src={groupIcon} alt="" />
        <Link to={{ pathname: `/group/${group.idgroup}`, state: { groupId } }} className="GP-title"><div className="GP-title">{group.name}</div></Link>
        {userId !== 0 && (
        <button type="submit" className={isJoined ? "GP-btn-joined" : "GP-btn-join"} intent="primary" onClick={unJoin}>
          {buttonText}
          {" "}
        </button>
        )}
      </div>
      <div className="GP-hashTags">
        {hashTagsList}
      </div>
      <div className="GP-description">
        {group.description}
      </div>
      <hr className="GP-horizontal-divider" />
      <div className="GP-footer">
        <span>
          {userNum}
          {" "}
          Members
        </span>
        <Divider />
        <span>
          {postNum}
          {" "}
          Posts
        </span>
      </div>
    </div>
  );
};

export default GroupPanel;
