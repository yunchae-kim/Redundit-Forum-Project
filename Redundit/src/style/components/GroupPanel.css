.GroupPanel-small {
    width: 500px;
    height: 300px;
    background-color: white;
    border-radius: 20px;
    color: black;
    font-size: 20px;
    margin-bottom: 1rem;
}

.GroupPanel-large {
  width: 750px;
  height: 300px;
  background-color: white;
  border-radius: 20px;
  color: black;
  font-size: 20px;
}

.GP-header {
  display: flex;
  margin-left: 28px;
  margin-right: 28px;
  padding-top: 16px;
  padding-bottom: 16px;
}

.GP-btn-join {
  background-color: #188bd8;
  color: white;
  margin-left: auto;
  margin-top: 0;
  width: 120px;
  height: 40px;
  border-radius: 20px;
  align-self: center;
  box-shadow: none;
  font-size: 20px;
  border: none;
  font-weight:600;
}

.GP-btn-joined {
  background-color:#a3d1ef;
  color: white;
  margin-left: auto;
  margin-top: 0;
  width: 120px;
  height: 40px;
  border-radius: 20px;
  align-self: center;
  box-shadow: none;
  font-size: 20px;
  border:none;
  font-weight:600;
}

.GP-title {
  align-self: center;
  margin-right: auto;
  font-size: 24px;
  font-weight: 500;
  color: #111;
}

.GP-groupIcon {
  border-radius: 50%;
  width: 40px;
  height: 40px;
  align-self: center;
  margin-right: 24px;
}

.GP-hashTags{
  display: flex;
  justify-content: flex-start;
  height: 1.5rem;
  margin-left: 28px;
  margin-right: 28px;
  margin-bottom: 24px;
}

.GP-hashTags-large{
  display: flex;
  justify-content: flex-start;
  margin-left: 27px;
  margin-right: 32px;
  padding-bottom: 15px;
}

span.bp3-tag.GP-tag{
  font-size: 16px;
}

.GP-tag {
  width: 100px;
  height: 32px;
  border-radius: 20px;
  margin-right: 10px;
}

.GP-description {
  margin-left: 28px;
  margin-right: 28px;
  padding-bottom: 15px;
  height: 90px;
  text-align: left;
  overflow: scroll;
}

.GP-footer {
  margin-left: 28px;
  margin-right: 28px;
  display: flex;
  justify-content: space-around;
  font-size: 16px;
  padding-top: 1rem;
  font-weight:600;
  color: #555555;
}

.GP-horizontal-divider {
  background-color: #dddddd;
  border:none;
  height:1px;
}