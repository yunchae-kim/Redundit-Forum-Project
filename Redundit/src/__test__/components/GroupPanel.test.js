import {
  waitFor,
} from "@testing-library/react";
import renderer from "react-test-renderer";
import "@testing-library/jest-dom";
import { BrowserRouter } from "react-router-dom";
import GroupPanel from "../../components/GroupPanel";

test("test should GroupPanel match snapshot", async () => {
  const tree = renderer.create(
    <BrowserRouter>
      <GroupPanel groupId={1} />
    </BrowserRouter>,
  ).toJSON();
  await waitFor(async () => {
    expect(tree).toMatchSnapshot();
  });
});
