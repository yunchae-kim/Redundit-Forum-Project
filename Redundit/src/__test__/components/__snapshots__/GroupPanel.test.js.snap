// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test should GroupPanel match snapshot 1`] = `
<div
  className="GroupPanel GroupPanel-small"
>
  <div
    className="GP-header"
  >
    <img
      alt=""
      className="GP-groupIcon"
      src="groupIcon.jpg"
    />
    <a
      className="GP-title"
      href="/group/undefined"
      onClick={[Function]}
    >
      <div
        className="GP-title"
      />
    </a>
  </div>
  <div
    className="GP-hashTags"
  />
  <div
    className="GP-description"
  />
  <hr
    className="GP-horizontal-divider"
  />
  <div
    className="GP-footer"
  >
    <span>
      0
       
      Members
    </span>
    <div
      className="bp3-divider"
    />
    <span>
      0
       
      Posts
    </span>
  </div>
</div>
`;
